src/backend/dx11/src/device.rs

```diff
@@ -1320,50 +1494,138 @@ impl hal::Device<Backend> for Device {
         J: IntoIterator,
         J::Item: Borrow<Sampler>,
     {
-        let mut max_binding = 0;
         let mut bindings = Vec::new();
 
-        // convert from DescriptorSetLayoutBinding to our own PipelineBinding, and find the higher
-        // binding number in the layout
+        let mut register_remap = Vec::new();
+        let mut num_t = 0;
+        let mut num_s = 0;
+        let mut num_c = 0;
+        let mut num_u = 0;
+
+        // we check how many hlsl registers we should use
         for binding in layout_bindings {
             let binding = binding.borrow();
 
-            max_binding = max_binding.max(binding.binding as u32);
-
-            bindings.push(PipelineBinding {
-                stage: binding.stage_flags,
-                ty: binding.ty,
-                binding_range: binding.binding..(binding.binding + 1),
-                handle_offset: 0
-            });
+            let hlsl_reg = match binding.ty {
+                pso::DescriptorType::Sampler => {
+                    num_s += 1;
+                    num_s
+                }
+                pso::DescriptorType::CombinedImageSampler => {
+                    num_t += 1;
+                    num_s += 1;
+                    num_t
+                }
+                pso::DescriptorType::SampledImage | pso::DescriptorType::UniformTexelBuffer => {
+                    num_t += 1;
+                    num_t
+                }
+                pso::DescriptorType::UniformBuffer | pso::DescriptorType::UniformBufferDynamic => {
+                    num_c += 1;
+                    num_c
+                }
+                pso::DescriptorType::StorageTexelBuffer | pso::DescriptorType::StorageBuffer |
+                pso::DescriptorType::InputAttachment | pso::DescriptorType::StorageBufferDynamic |
+                pso::DescriptorType::StorageImage => {
+                    num_u += 1;
+                    num_u
+                }
+            } - 1;
+
+            // we decompose combined image samplers into a separate sampler and image internally
+            if binding.ty == pso::DescriptorType::CombinedImageSampler {
+                // TODO: for now we have to make combined image samplers share registers since
+                //       spirv-cross doesn't support setting the register of the sampler/texture
+                //       pair to separate values (only one `DescriptorSet` decorator)
+                let shared_reg = num_s.min(num_t);
+
+                num_s = shared_reg;
+                num_t = shared_reg;
```

Comment:
wouldn't this break any following images/sampler? we just reset the register number to a technically unrelated value, and we are still going to increment this and use it afterwards?

Reply:
`shared_reg` should be using `max()` 😓

Ideally we shouldn't have to do this, but because of spirv-cross limitations we potentially end up with some wasted space, see image:
![register allocation](https://i.imgur.com/yO9YjTr.png)